### ISSUE TITLE
[RFC] Allow different default version contraints when using `composer require`

### ISSUE THREAD

**Issue by benjaminmal** [composer/composer#10581]
Currently, when we use `composer require vendor/package` the version contraints is automatically a "bump" (e.g. `^1.0`). It would be nice to have the possibility to add an other default version contraint. It's really annoying to either find the version in packagist and set it in the `composer require vendor/package:version-contraints` or doing it in the `composer.json` after a `composer require vendor/package` if you have a different version strategy from the default behavior.

Following the [docs](https://getcomposer.org/doc/articles/versions.md#writing-version-constraints) I see 4 versions contraints we could implement:
- [Exact Version Constraint](https://getcomposer.org/doc/articles/versions.md#exact-version-constraint): 
    - `exact`: `1.0.0`
- [Caret Version Range](https://getcomposer.org/doc/articles/versions.md#caret-version-range-): 
    - `caret`: `^1.0` <--- ***Still remains the default version constraints***
- [Wildcard Version Range](https://getcomposer.org/doc/articles/versions.md#wildcard-version-range-): 
    - `wildcard-major`: `*`
    - `wildcard-minor`: `1.*`
    - `wildcard-patch`: `1.0.*`
- [Tilde Version Range](https://getcomposer.org/doc/articles/versions.md#tilde-version-range-): 
    - `tilde-major`: `~1` 
    - `tilde-minor`: `~1.0`
    - `tilde-patch`: `~1.0.0`

[Version Range](https://getcomposer.org/doc/articles/versions.md#version-range), [Hyphenated Version Range](https://getcomposer.org/doc/articles/versions.md#hyphenated-version-range-) and [Stability Constraints](https://getcomposer.org/doc/articles/versions.md#stability-constraints) contraints have been voluntarily ommited since I think they are edge-case version constraints.

Maybe something like that:

The `composer.json`:

```json
{
    "name": "vendor/project",
    "config": {
        "version-strategy": "exact"
    }
}
```

When I run this command:

```bash
composer require vendor/package
```

I get the following composer.json:

```json
{
    "name": "vendor/project",
    "config": {
        "version-strategy": "exact"
    },
    "require": {
        "vendor/package": "^1.0"
    }
}
```

And I expected this to happen:
```json
{
    "name": "vendor/project",
    "config": {
        "version-strategy": "exact"
    },
    "require": {
        "vendor/package": "1.0.0"
    }
}
```

Also, it would be nice to support it in CLI:
```bash
composer require vendor/package --version-strategy=exact
```

Or maybe is there already a way to do so?

**Comment by Seldaek** [comment]
What's the point of exact requires when you have the lock file to pin packages to an exact version? I don't really see the value in this, so not so inclined to bear the burden of maintenance for such a feature.

**Comment by benjaminmal** [comment]
It's an old debate I guess. You can see [this article](https://docs.renovatebot.com/dependency-pinning/#pinning-dependencies-and-lock-files) explaining that pinning dependencies and have a lock file may have a meaning for some maintainers. 

I can see other reasons:
- You create a test repo for a bug fix, so you need to pin them all to be able to reproduce the bug
- For some reason you want to pin dependencies for a library (without composer.lock), because you are dependent of a package known for BC breaks
- For a project I use with renovate, the bot was supposed to update a patch and automerge but instead it updates a minor which break the build (it was a bug from renovate). If I had pinned my dependencies, composer wouldn't allowing it. Now I pinned my dependencies, I am sure that renovate would not break things.
- If for some reason the composer.lock file is broken (e.g. a bad merge conflict), you can lost all your pinned dependencies and it can be a pain to have them back or to update your legacy code.

But the proposal is not only about `exact` version strategy or should I pin/bump all dependencies, it's about giving the developer the opportunity to use a different version strategy out of the box. Either in CLI or as a default strategy.

Not everyone needs `^1.0` as a default strategy, even if the majority would 😄

**Comment by benjaminmal** [comment]
And also, composer already allows to pin dependencies. It would simply be easier to do so. If this proposal is accepted, it would have no sense to not included `exact` version strategy.

**Comment by Seldaek** [comment]
I could see the following strategies are valid, the others you list above IMO do not make much practical sense:

- `fixed` / `exact` / `pinned`: `1.2.3` always
- `semver`: `^1.2` or equivalent, current behavior
- `patch-only`: `~1.2.3` 

That said, there doesn't seem to be a ton of demand for this, so I am a bit reluctant to add yet another config option.. but note that you can already do `composer require foo/bar --fixed` to get an exact version added. IMO this makes more sense to do it on demand anyway, as having it set by default is probably a bad idea given most packages follow semver, and you have the lock file too.
